### ISSUE TITLE
@AddCSRFToken sets underlying request to null

### ISSUE THREAD

**Issue by itsbrnrd** [playframework/playframework#6189]
### Environment
* Play! 2.5.3
* java version "1.8.0_72"
* OSX 10.10

### Issue
I have a strange behaviour when using `AddCSRFToken` annotation.

The following code snippet:
``` java
@AddCSRFToken
public Result ping() {
    DynamicForm form = this.formFactory.form().bindFromRequest();
    return ok(CSRF.getToken(request()).map(CSRF.Token::value).orElse("no token"));
}
```
throws a NPE on `bindFromRequest()` that comes from `Form#requestData(Http.Request request)` when it tries to read the request's body.

After some investigation it seems that when `AddCSRFTokenAction` is called it sets a new `Http.Context` with a new `RequestImpl` based on the current RequestHeader containing the CSRF tags and a null underlying `Request` (https://github.com/playframework/playframework/blob/master/framework/src/play-filters-helpers/src/main/java/play/filters/csrf/AddCSRFTokenAction.java#L63-L73).

**Comment by gmethvin** [comment]
@itsbrnrd Can you clarify where you see it getting set to null? From what I can tell it's setting it to the new tagged request. Also it would be helpful if you can provide the exception stack trace you're getting.

**Comment by itsbrnrd** [comment]
Sure. Here is the complete exception stack trace:

```
java.util.concurrent.CompletionException: java.lang.NullPointerException
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292)
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:593)
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977)
	at scala.concurrent.java8.FuturesConvertersImpl$CF.apply(FutureConvertersImpl.scala:21)
	at scala.concurrent.java8.FuturesConvertersImpl$CF.apply(FutureConvertersImpl.scala:18)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
	at scala.concurrent.BatchingExecutor$Batch$$anonfun$run$1.processBatch$1(BatchingExecutor.scala:63)
	at scala.concurrent.BatchingExecutor$Batch$$anonfun$run$1.apply$mcV$sp(BatchingExecutor.scala:78)
	at scala.concurrent.BatchingExecutor$Batch$$anonfun$run$1.apply(BatchingExecutor.scala:55)
	at scala.concurrent.BatchingExecutor$Batch$$anonfun$run$1.apply(BatchingExecutor.scala:55)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:72)
	at scala.concurrent.BatchingExecutor$Batch.run(BatchingExecutor.scala:54)
	at scala.concurrent.Future$InternalCallbackExecutor$.unbatchedExecute(Future.scala:599)
	at scala.concurrent.BatchingExecutor$class.execute(BatchingExecutor.scala:106)
	at scala.concurrent.Future$InternalCallbackExecutor$.execute(Future.scala:597)
	at scala.concurrent.impl.CallbackRunnable.executeWithValue(Promise.scala:40)
	at scala.concurrent.impl.Promise$DefaultPromise.tryComplete(Promise.scala:248)
	at scala.concurrent.Promise$class.complete(Promise.scala:55)
	at scala.concurrent.impl.Promise$DefaultPromise.complete(Promise.scala:153)
	at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:235)
	at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:235)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
	at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)
	at akka.dispatch.BatchingExecutor$BlockableBatch$$anonfun$run$1.apply$mcV$sp(BatchingExecutor.scala:91)
	at akka.dispatch.BatchingExecutor$BlockableBatch$$anonfun$run$1.apply(BatchingExecutor.scala:91)
	at akka.dispatch.BatchingExecutor$BlockableBatch$$anonfun$run$1.apply(BatchingExecutor.scala:91)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:72)
	at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:90)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:39)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:405)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: java.lang.NullPointerException: null
	at play.data.Form.requestData(Form.java:181)
	at play.data.DynamicForm.bindFromRequest(DynamicForm.java:76)
	at controllers.Test.ping(Test.java:103)
	at router.Routes$$anonfun$routes$1$$anonfun$applyOrElse$26$$anonfun$apply$26.apply(Routes.scala:723)
	at router.Routes$$anonfun$routes$1$$anonfun$applyOrElse$26$$anonfun$apply$26.apply(Routes.scala:723)
	at play.core.routing.HandlerInvokerFactory$$anon$4.resultCall(HandlerInvoker.scala:157)
	at play.core.routing.HandlerInvokerFactory$$anon$4.resultCall(HandlerInvoker.scala:156)
	at play.core.routing.HandlerInvokerFactory$JavaActionInvokerFactory$$anon$14$$anon$3$$anon$1.invocation(HandlerInvoker.scala:136)
	at play.core.j.JavaAction$$anon$1.call(JavaAction.scala:73)
	at play.http.HttpRequestHandler$1.call(HttpRequestHandler.java:54)
	at play.filters.csrf.AddCSRFTokenAction.call(AddCSRFTokenAction.java:76)
	at play.core.j.JavaAction$$anonfun$7.apply(JavaAction.scala:108)
	at play.core.j.JavaAction$$anonfun$7.apply(JavaAction.scala:108)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
	at play.core.j.HttpExecutionContext$$anon$2.run(HttpExecutionContext.scala:56)
	at play.api.libs.iteratee.Execution$trampoline$.execute(Execution.scala:70)
	at play.core.j.HttpExecutionContext.execute(HttpExecutionContext.scala:48)
	at scala.concurrent.impl.Future$.apply(Future.scala:31)
	at scala.concurrent.Future$.apply(Future.scala:492)
	at play.core.j.JavaAction.apply(JavaAction.scala:108)
	at play.api.mvc.Action$$anonfun$apply$2$$anonfun$apply$5$$anonfun$apply$6.apply(Action.scala:112)
	at play.api.mvc.Action$$anonfun$apply$2$$anonfun$apply$5$$anonfun$apply$6.apply(Action.scala:112)
	at play.utils.Threads$.withContextClassLoader(Threads.scala:21)
	at play.api.mvc.Action$$anonfun$apply$2$$anonfun$apply$5.apply(Action.scala:111)
	at play.api.mvc.Action$$anonfun$apply$2$$anonfun$apply$5.apply(Action.scala:110)
	at scala.Option.map(Option.scala:146)
	at play.api.mvc.Action$$anonfun$apply$2.apply(Action.scala:110)
	at play.api.mvc.Action$$anonfun$apply$2.apply(Action.scala:103)
	at scala.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:251)
	at scala.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:249)
	... 13 common frames omitted
```

**Comment by gmethvin** [comment]
Okay. Thanks. Now that I look at it the issue is actually that we're constructing a `RequestImpl` from a `RequestHeader` rather than the full `Request`.

**Comment by itsbrnrd** [comment]
Not sure if I'm really clear so here are screenshots of the debugger inside the annotation after the new context is created.
<img width="1024" alt="screen shot 2016-05-25 at 10 49 13" src="https://cloud.githubusercontent.com/assets/5990900/15533764/60cdbc68-2266-11e6-9feb-e4ea2a7b9a0f.png">

<img width="1043" alt="screen shot 2016-05-25 at 10 49 00" src="https://cloud.githubusercontent.com/assets/5990900/15533765/60cf1b1c-2266-11e6-872d-baa35241885d.png">

When calling `request()` `underlying` is null whereas when accessing the `request` field it's not.

**Comment by gmethvin** [comment]
Yeah, the constructor that takes the `RequestHeader` (used here) explicitly sets underlying to null: https://github.com/playframework/playframework/blob/2.5.3/framework/src/play/src/main/java/play/mvc/Http.java#L678. In this case we should just be constructing a full Scala Request in the filter instead, and use the constructor that takes a full request.

I also think the API for `RequestImpl` is a bit confusing. We should fix that too.

**Comment by itsbrnrd** [comment]
I'm no Scala developer but I'd be happy to contribute if fixing this issue doesn't require to rewrite a full Scala class.

Any hints or recommendations in the way we should fix this issue?

**Comment by gmethvin** [comment]
That's fine. I was already familiar with that code so I submitted a fix: #6192

**Comment by itsbrnrd** [comment]
👍 Thanks!
